### PR TITLE
Research implementation.

### LINKED ISSUE #2: Explore: extracting research data from EXE
Explore possibility of extracting research data from game EXE.

Strings are in string table at the end of EXE, so a hypothetical table with research data will be more difficult to spot.

Name of researched technologies:
```
Plasma weapons
Plasma Bomb
Light Plasma Canon
Plasma Canon
Plasma fusion
Nuclear weapons
Ion canon
Light Laser
Laser Cannon
Laser detector
Teleportation
AI mine
Heavy rockets
Rocket engine
Engine
Flying
Fast Repairing
High Medicine
Nuclear engine
Nuclear reactor
```
Names from E2140.exe from Steam release, SHA256: c7230a9348047093437aeb1b82e3c222b15fb24d5318468247cc09d9517646e9

### LINKED ISSUE #471: Technology research

### PR DESCRIPTION
**Pull request by IceReaper**
Fully implemented research system. Closes #471

### CODE REVIEW

**Comment by Dzierzan** [comment]
Works great!

**Comment by michaeldgg2** [comment]
Results from testing:
- research tab is enabled before the construction of Research Center is finished
- alignment issues in the UI
img
- couldn't test research conquering

File: OpenRA.Mods.OpenE2140/Widgets/Logic/ResearchLogic.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+		this.UpdateBackgrounds(this.production.DisplayedIconCount, this.research.Researchables.Length);
+	}
+
+	private void SwitchToProduction()
+	{
+		this.research.Visible = false;
+		this.production.Visible = true;
+
+		this.UpdateBackgrounds(this.research.Researchables.Length, this.production.DisplayedIconCount);
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Widgets/Logic/ResearchLogic.cs:94]: This is still called for every tick, right? Shouldn't the check for when to invoke `OnIconCountChanged` event be more robust and done for both research and production tabs?

Comment (IceReaper) [reply OpenRA.Mods.OpenE2140/Widgets/Logic/ResearchLogic.cs:94]: No its not. Once when switching to research, once when switching back, and once when the amount of researchables changed.

**Review by michaeldgg2 — APPROVED** [review]

**Pull request merged**